File: adventofcode/2020/day18/day18.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
)

const (
	ADD  = '+'
	MUL  = '*'
	PARL = '('
	PARR = ')'
)

func precedence(r rune) int {
	switch r {
	case ADD:
		return 2
	case MUL:
		return 1
	}
	return 0
}

func evalOp(a, b int, op rune) int {
	switch op {
	case ADD:
		return a + b
	case MUL:
		return a * b
	}
	return 0
}

func evalOpOnStack(ops *Stack, vals *Stack) {
	val2 := vals.pop()
	val1 := vals.pop()
	op := ops.pop()

	vals.push(evalOp(val1, val2, rune(op)))
}

type Stack struct {
	values []int
}

func (s *Stack) empty() bool {
	return len(s.values) == 0
}

func (s *Stack) top() int {
	return s.values[len(s.values)-1]
}

func (s *Stack) pop() int {
	top := s.top()
	s.values = s.values[:len(s.values)-1]
	return top
}

func (s *Stack) push(a int) {
	s.values = append(s.values, a)
}

var digitRegexp = regexp.MustCompile(`^\d$`)

func isDigit(r byte) bool {
	return digitRegexp.MatchString(string(r))
}

func evaluate(tokens string) int {
	ops := &Stack{}
	vals := &Stack{}

	for i := 0; i < len(tokens); i++ {
		if tokens[i] == ' ' {
			continue
		}

		if isDigit(tokens[i]) {
			num := 0
			for ; i < len(tokens) && isDigit(tokens[i]); i++ {
				num = num*10 + int(tokens[i]-'0')
			}
			vals.push(num)
			i--
		}

		if tokens[i] == PARL {
			ops.push(int(tokens[i]))
		}

		if tokens[i] == PARR {
			for !ops.empty() && ops.top() != PARL {
				evalOpOnStack(ops, vals)
			}

			if !ops.empty() {
				ops.pop()
			}
		}

		if tokens[i] == ADD || tokens[i] == MUL {
			for !ops.empty() && precedence(rune(ops.top())) >= precedence(rune(tokens[i])) {
				evalOpOnStack(ops, vals)
			}

			ops.push(int(tokens[i]))
		}
	}

	for !ops.empty() {
		evalOpOnStack(ops, vals)
	}

	return vals.top()
}

func readInput() []string {
	lines := []string{}
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines
}

func main() {
	exprs := readInput()
	sum := 0
	for _, expr := range exprs {
		sum += evaluate(expr)
	}
	fmt.Println(sum)
}

File: adventofcode/2020/day18/day18.cpp
#include<iostream>
#include<string>
#include<stack>
#include<ctype.h>
using namespace std;

const char ADD = '+';
const char MUL = '*';
const char PARL = '(';
const char PARR = ')';

int precedence(char c) {
  switch (c) {
    case ADD: return 2;
    case MUL: return 1;
  }
  return 0;
}

long int evalOp(long int a, long int b, char op) {
  switch (op) {
    case ADD: return a + b;
    case MUL: return a * b;
  }
  return 0;
}

void evalOpOnStack(stack<char> &ops, stack<long int> &vals) {
  long int val2 = vals.top();
  vals.pop();

  long int val1 = vals.top();
  vals.pop();

  char op = ops.top();
  ops.pop();

  vals.push(evalOp(val1, val2, op));
}

long int evaluate(string expr) {
  stack<char> ops;
  stack<long int> vals;

  for (int i = 0; i < expr.length(); i++) {
    if (expr[i] == ' ')
      continue;

    else if (isdigit(expr[i])) {
      int val = 0;
      for (; i < expr.length() && isdigit(expr[i]); i++) {
        val = val*10 + (expr[i]-'0');
      }
      vals.push(val);
      i--;
    }

    else if (expr[i] == PARL) {
      ops.push(expr[i]);
    }

    else if (expr[i] == PARR) {
      while (!ops.empty() && ops.top() != PARL) {
        evalOpOnStack(ops, vals);
      }

      if (!ops.empty()) {
        ops.pop();
      }
    }

    else {
      while (!ops.empty() && precedence(ops.top()) >= precedence(expr[i])) {
        evalOpOnStack(ops, vals);
      }

      ops.push(expr[i]);
    }
  }

  while (!ops.empty()) {
    evalOpOnStack(ops, vals);
  }

  return vals.top();
}

int main() {
  long int sum = 0;
  for (string expr; getline(cin, expr);) {
    sum += evaluate(expr);
  }
  cout << sum << endl;
  return 0;
}
